### PR TITLE
Implement cards v3 (2nd September 2016)

### PR DESCRIPTION
**Pull request by Patashu**
taking into account all feedback and code changes

### CODE REVIEW

File: Brimstone/Actions.cs [hunk 1]
```diff
@@ -332,10 +332,156 @@ public class Damage : QueueAction
...
+		public const int TARGETS = 0;
+
+		public override ActionResult Run(Game game, IEntity source, List<ActionResult> args) {
+			throw new NotImplementedException(); // TODO: implement https://hearthstone.gamepedia.com/Advanced_rulebook#Destroy_effects_in_all_zones
+		}
+	}
+
+	public class GainMana : QueueAction
+	{
+		public const int CONTROLLER = 0;
```
Comment (djkaty) [inline Brimstone/Actions.cs]: PLAYER, apparently, since that's what I've used in Give and Draw it seems.

File: Brimstone/Actions.cs [hunk 2]
```diff
@@ -332,10 +332,156 @@ public class Damage : QueueAction
...
+		public const int AMOUNT = 1;
+
+		public override ActionResult Run(Game game, IEntity source, List<ActionResult> args) {
+			throw new NotImplementedException(); // TODO: implement https://hearthstone.gamepedia.com/Advanced_rulebook#Mana_Crystals_and_mana_costs
+		}
+	}
+
+	public class Summon : QueueAction
+	{
+		public const int CONTROLLER = 0;
```
Comment (djkaty) [inline Brimstone/Actions.cs]: PLAYER

File: Brimstone/Actions.cs [hunk 3]
```diff
@@ -332,10 +332,156 @@ public class Damage : QueueAction
...
+			if (args[TARGETS].HasResult)
+				foreach (Hero e in args[TARGETS]) {
+					DebugLog.WriteLine("{0} gains {1} Armor", e.ShortDescription, args[AMOUNT]);
+					e.GainArmour(args[AMOUNT]);
+				}
+			return ActionResult.None;
+		}
+	}
+
+	public class EquipWeapon : QueueAction
```
Comment (djkaty) [inline Brimstone/Actions.cs]: So we can get rid of this or?

File: Brimstone/Actions.cs [hunk 4]
```diff
@@ -332,10 +332,156 @@ public class Damage : QueueAction
...
+	{
+		public const int MINION1 = 0;
+		public const int MINION2 = 1;
+
+		public override ActionResult Run(Game game, IEntity source, List<ActionResult> args) {
+			throw new NotImplementedException(); // TODO: implement https://www.youtube.com/watch?v=YhJlDd7NxNg + log in description
+		}
+	}
+
+	public class GiveTaunt : QueueAction
```
Comment (djkaty) [inline Brimstone/Actions.cs]: Kill this one

File: Brimstone/Actions.cs [hunk 5]
```diff
@@ -332,10 +332,156 @@ public class Damage : QueueAction
...
+				foreach (ICharacter e in args[TARGETS]) {
+					DebugLog.WriteLine("{0} is given Taunt", e.ShortDescription);
+
+					e.HasTaunt = true;
+				}
+			return ActionResult.None;
+		}
+	}
+
+	public class GainArmour : QueueAction
```
Comment (djkaty) [inline Brimstone/Actions.cs]: Kill this one

File: Brimstone/Behaviours.cs [hunk 6]
```diff
@@ -21,24 +21,65 @@ public class Actions {
...
+		// Random minion/character - all
+		public static ActionGraph RandomMinion { get { return Random(AllMinions); } }
+		public static ActionGraph RandomFriendlyMinion { get { return Random(FriendlyMinions); } }
 		public static ActionGraph RandomOpponentMinion { get { return Random(OpponentMinions); } }
+		public static ActionGraph RandomCharacter { get { return Random(AllCharacters); } }
+		public static ActionGraph RandomFriendlyCharacter { get { return Random(FriendlyCharacters); } }
 		public static ActionGraph RandomOpponentCharacter { get { return Random(OpponentCharacters); } }
+		public static ActionGraph RandomFriendlyMinionInDeck { get { return Random(FriendlyMinionsInDeck); } }
+		public static ActionGraph RandomOpponentMinionInDeck { get { return Random(OpponentMinionsInDeck); } }
+		public static ActionGraph RandomFriendlyMinionInHand { get { return Random(FriendlyMinionsInHand); } }
```
Comment (djkaty) [inline Brimstone/Behaviours.cs]: Can you just separate out the *Healthy* ones into a group underneath the original ones, with a newline

File: Brimstone/Cards/All.cs [hunk 7]
```diff
@@ -1,47 +1,424 @@
...
+		// TODO: Enrage, Enchantments, gaining Windfury (EX1_412e)
+
+		// Murloc Tidehunter
+		public static Behaviour EX1_506 = new Behaviour {
+			Battlecry = Summon(Controller, "EX1_506a")
+		};
+
+		// Harvest Golem
+		public static Behaviour EX1_556 = new Behaviour {
+			Deathrattle = Summon(Controller, "skele21")
```
Comment (djkaty) [inline Brimstone/Cards/All.cs]: Do the murloc from Tidehunter and the Damaged Golem have unique names? If so, please use them for readability :) I'm planning to make the reflector also find behaviour class names like 'HarvestGolem' in future too

File: Brimstone/Cards/All.cs [hunk 8]
```diff
@@ -1,47 +1,424 @@
...
+			Battlecry = (Draw(Controller) * 2).Then((Draw(Opponent) * 2))
+		};
+
+		// Mana Addict
+		// public static Behaviour EX1_055 = new Behaviour {
+		// TODO: Conditional selectors, Enchantments (EX1_055o)
+
+		// Sunfury Protector
+		public static Behaviour EX1_058 = new Behaviour {
+			Battlecry = GiveTaunt(AdjacentMinions)
```
Comment (djkaty) [inline Brimstone/Cards/All.cs]: Comment this and similar out til we figure out the tag setting action

File: Brimstone/Cards/All.cs [hunk 9]
```diff
@@ -1,47 +1,424 @@
...
+		// TODO: Combining selectors, Enchantments (EX1_509e)
+
+		// Ancient Mage
+		// public static Behaviour EX1_584 = new Behaviour {
+		// TODO: Enchantments (EX1_584e)
+
+		// Imp Master
+		public static Behaviour EX1_597 = new Behaviour {
+			Triggers = {
+				OnEndTurn(Controller, Damage(Self, 1).Then(Summon(Controller, "EX1_598")))
```
Comment (djkaty) [inline Brimstone/Cards/All.cs]: See above re: card names

File: Brimstone/Cards/All.cs [hunk 10]
```diff
@@ -1,47 +1,424 @@
...
 		};
 
+		// Arcane Explosion
+		public static Behaviour CS2_025 = new Behaviour {
+			Battlecry = Damage(OpponentMinions, 1)
+		};
+
+		// Moonfire
+		public static Behaviour CS2_008 = new Behaviour {
+			Battlecry = Damage(Target, 1)
```
Comment (djkaty) [inline Brimstone/Cards/All.cs]: We need to write code to determine if the damage is comingfrom a spell and apply spell damage, can you add a TODO note for that

File: Brimstone/Cards/All.cs [hunk 11]
```diff
@@ -1,47 +1,424 @@
...
+		public static Behaviour EX1_402 = new Behaviour {
+			Triggers = {
+				OnDamage(FriendlyMinions, GainArmour(FriendlyHero, 1))
+			}
+		};
+
 		// Imp Gang Boss
+		public static Behaviour BRM_006 = new Behaviour {
+			Triggers = {
+				OnDamage(Self, Summon(Controller, "BRM_006t"))
```
Comment (djkaty) [inline Brimstone/Cards/All.cs]: Above note re: card names

File: Brimstone/Cards/All.cs [hunk 12]
```diff
@@ -1,47 +1,424 @@
...
+			}
+		};
+
+		// Arathi Weaponsmith
+		public static Behaviour EX1_398 = new Behaviour {
+			Battlecry = EquipWeapon(Controller, "EX1_398t")
```
Comment (djkaty) [inline Brimstone/Cards/All.cs]: Above note re: card names

File: Brimstone/Cards/All.cs [hunk 13]
```diff
@@ -1,47 +1,424 @@
...
+
+		//Deathlord
+		public static Behaviour FP1_009 = new Behaviour {
+			Deathrattle = Summon(Opponent, RandomOpponentMinionInDeck, 1)
+		};
+
+		//Illidan Stormrage
+		public static Behaviour EX1_614 = new Behaviour {
+			Triggers = {
+				OnPlay(Controller, Summon(Controller, "EX1_614t"))
```
Comment (djkaty) [inline Brimstone/Cards/All.cs]: Above note re: card names

File: Brimstone/EntityTypes/Hero.cs [hunk 14]
```diff
@@ -17,5 +17,9 @@ public class Hero : Character<Hero>
 		public override object Clone() {
 			return new Hero(this);
 		}
+
+		public void GainArmour(int amount) {
+			this[GameTag.ARMOR] += amount;
```
Comment (djkaty) [inline Brimstone/EntityTypes/Hero.cs]: Okay, semantics guide. Code in the entities should never EVER work like this. If a player is writing a game script, they use the entity methods as a simplified API to the action queue. Check out the methods in Player.cs etc. Calling a method on an entity always just queues something up. Similarly, this should queue up GainArmor (or the tag change action we haven't written yet). The Actions should generally be the only thing that changes the tags. Also, when you do change tags, you should generally use the properties in TagProperties.cs to do so unless you're writing internal code (for example in Zone.cs), so you would do Armor += amount here. This saves on typecasting sometimes and reads more nicely, but more importantly it gives us additional hooks to do other things later, like for example we might set 'LastPlayerThatGainedArmor' or something via the tag property too.

In a nutshell: entity methods should queue actions. Tag changes should occur in actions and use the tag properties to do so. Since we're going to write a tag modifying function, I think you can just delete this code for now :)
